lib/src/wallet.rs

```diff
@@ -1198,77 +1215,165 @@ impl LightWallet {
+                    sapling_notes,
+                    utxos,
+                    (transparent_value_selected + sapling_value_selected + orchard_value_selected)
+                        .unwrap(),
                 );
             }
         }
 
         // If we can't select enough, then we need to return empty handed
-        (vec![], vec![], Amount::zero())
+        (vec![], vec![], vec![], Amount::zero())
+    }
+
+    async fn get_all_domain_specific_notes<D>(&self) -> Vec<Vec<D::SpendableNoteAT>>
+    where
+        D: DomainWalletExt<zingoconfig::Network>,
+        <D as Domain>::Recipient: traits::Recipient,
+        <D as Domain>::Note: PartialEq + Clone,
+    {
+        let keys_arc = self.keys();
+        let keys = keys_arc.read().await;
+        let notes_arc = self.transactions();
+        let notes = notes_arc.read().await;
+        self.transaction_context
+            .config
+            .anchor_offset
+            .iter()
+            .map(|anchor_offset| {
+                let mut candidate_notes = notes
+                    .current
+                    .iter()
+                    .flat_map(|(transaction_id, transaction)| {
+                        D::WalletNote::transaction_metadata_notes(transaction)
+                            .iter()
+                            .map(move |note| (*transaction_id, note))
+                    })
+                    .filter(|(_, note)| note.value() > 0)
+                    .filter_map(|(transaction_id, note)| {
+                        // Filter out notes that are already spent
+                        if note.spent().is_some() || note.unconfirmed_spent().is_some() {
+                            None
+                        } else {
+                            // Get the spending key for the selected fvk, if we have it
+                            let extsk = keys.get_spend_key_for_fvk::<D>(&note.fvk());
+                            SpendableNote::from(
+                                transaction_id,
+                                note,
+                                *anchor_offset as usize,
+                                &extsk,
+                            )
+                        }
+                    })
+                    .collect::<Vec<D::SpendableNoteAT>>();
+                candidate_notes.sort_by_key(|spendable_note| {
+                    D::WalletNote::value_from_note(&spendable_note.note())
+                });
+                candidate_notes
+            })
+            .collect()
+    }
+
+    fn add_notes_to_total<D: DomainWalletExt<zingoconfig::Network>>(
+        candidates: Vec<Vec<D::SpendableNoteAT>>,
+        target_amount: Amount,
+    ) -> (Vec<D::SpendableNoteAT>, Amount)
+    where
+        D::Note: PartialEq + Clone,
+        D::Recipient: traits::Recipient,
+    {
+        let mut notes = vec![];
+        let mut value_selected = Amount::zero();
+        let mut candidates = candidates.into_iter();
+        loop {
```

Comment:
Why not a `for` loop?

Reply:
https://github.com/zingolabs/zingolib/issues/114#issuecomment-1259853805